File: TESTING/EIG/zhbt21.f
      SUBROUTINE ZHBT21( UPLO, N, KA, KS, A, LDA, D, E, U, LDU, WORK, RWORK, RESULT )

*  -- LAPACK test routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      String             UPLO;
      int                KA, KS, LDA, LDU, N;
      // ..
      // .. Array Arguments ..
      double             D( * ), E( * ), RESULT( 2 ), RWORK( * );
      COMPLEX*16         A( LDA, * ), U( LDU, * ), WORK( * )
      // ..

*  =====================================================================

      // .. Parameters ..
      COMPLEX*16         CZERO, CONE
      const              CZERO = ( 0.0D+0, 0.0D+0 ), CONE = ( 1.0D+0, 0.0D+0 ) ;
      double             ZERO, ONE;
      const              ZERO = 0.0D+0, ONE = 1.0D+0 ;
      // ..
      // .. Local Scalars ..
      bool               LOWER;
      String             CUPLO;
      int                IKA, J, JC, JR;
      double             ANORM, ULP, UNFL, WNORM;
      // ..
      // .. External Functions ..
      bool               LSAME;
      double             DLAMCH, ZLANGE, ZLANHB, ZLANHP;
      // EXTERNAL LSAME, DLAMCH, ZLANGE, ZLANHB, ZLANHP
      // ..
      // .. External Subroutines ..
      // EXTERNAL ZGEMM, ZHPR, ZHPR2
      // ..
      // .. Intrinsic Functions ..
      // INTRINSIC DBLE, DCMPLX, MAX, MIN
      // ..
      // .. Executable Statements ..

      // Constants

      RESULT( 1 ) = ZERO
      RESULT( 2 ) = ZERO
      IF( N.LE.0 ) RETURN

      IKA = MAX( 0, MIN( N-1, KA ) )

      if ( LSAME( UPLO, 'U' ) ) {
         LOWER = .FALSE.
         CUPLO = 'U'
      } else {
         LOWER = .TRUE.
         CUPLO = 'L'
      }

      UNFL = DLAMCH( 'Safe minimum' )
      ULP = DLAMCH( 'Epsilon' )*DLAMCH( 'Base' )

      // Some Error Checks

      // Do Test 1

      // Norm of A:

      ANORM = MAX( ZLANHB( '1', CUPLO, N, IKA, A, LDA, RWORK ), UNFL )

      // Compute error matrix:    Error = A - U S U**H

      // Copy A from SB to SP storage format.

      J = 0
      for (JC = 1; JC <= N; JC++) { // 50
         if ( LOWER ) {
            DO 10 JR = 1, MIN( IKA+1, N+1-JC )
               J = J + 1
               WORK( J ) = A( JR, JC )
            } // 10
            DO 20 JR = IKA + 2, N + 1 - JC
               J = J + 1
               WORK( J ) = ZERO
            } // 20
         } else {
            DO 30 JR = IKA + 2, JC
               J = J + 1
               WORK( J ) = ZERO
            } // 30
            DO 40 JR = MIN( IKA, JC-1 ), 0, -1
               J = J + 1
               WORK( J ) = A( IKA+1-JR, JC )
            } // 40
         }
      } // 50

      for (J = 1; J <= N; J++) { // 60
         zhpr(CUPLO, N, -D( J ), U( 1, J ), 1, WORK );
      } // 60

      if ( N.GT.1 .AND. KS.EQ.1 ) {
         DO 70 J = 1, N - 1
            zhpr2(CUPLO, N, -DCMPLX( E( J ) ), U( 1, J ), 1, U( 1, J+1 ), 1, WORK );
         } // 70
      }
      WNORM = ZLANHP( '1', CUPLO, N, WORK, RWORK )

      if ( ANORM.GT.WNORM ) {
         RESULT( 1 ) = ( WNORM / ANORM ) / ( N*ULP )
      } else {
         if ( ANORM.LT.ONE ) {
            RESULT( 1 ) = ( MIN( WNORM, N*ANORM ) / ANORM ) / ( N*ULP )
         } else {
            RESULT( 1 ) = MIN( WNORM / ANORM, DBLE( N ) ) / ( N*ULP )
         }
      }

      // Do Test 2

      // Compute  U U**H - I

      zgemm('N', 'C', N, N, N, CONE, U, LDU, U, LDU, CZERO, WORK, N );

      for (J = 1; J <= N; J++) { // 80
         WORK( ( N+1 )*( J-1 )+1 ) = WORK( ( N+1 )*( J-1 )+1 ) - CONE
      } // 80

      RESULT( 2 ) = MIN( ZLANGE( '1', N, N, WORK, N, RWORK ), DBLE( N ) ) / ( N*ULP )

      RETURN

      // End of ZHBT21

      }

File: test/eig/zhbt21.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:dart_lapack/src/blas/zgemm.dart';
import 'package:dart_lapack/src/blas/zhpr.dart';
import 'package:dart_lapack/src/blas/zhpr2.dart';
import 'package:dart_lapack/src/complex.dart';
import 'package:dart_lapack/src/install/dlamch.dart';
import 'package:dart_lapack/src/install/lsame.dart';
import 'package:dart_lapack/src/matrix.dart';
import 'package:dart_lapack/src/zlange.dart';
import 'package:dart_lapack/src/zlanhb.dart';
import 'package:dart_lapack/src/zlanhp.dart';

void zhbt21(
  final String UPLO,
  final int N,
  final int KA,
  final int KS,
  final Matrix<Complex> A_,
  final int LDA,
  final Array<double> D_,
  final Array<double> E_,
  final Matrix<Complex> U_,
  final int LDU,
  final Array<Complex> WORK_,
  final Array<double> RWORK_,
  final Array<double> RESULT_,
) {
// -- LAPACK test routine --
// -- LAPACK is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  final A = A_.having(ld: LDA);
  final U = U_.having(ld: LDU);
  final WORK = WORK_.having();
  final RWORK = RWORK_.having();
  final D = D_.having();
  final E = E_.having();
  final RESULT = RESULT_.having(length: 2);
  const ZERO = 0.0, ONE = 1.0;
  bool LOWER;
  String CUPLO;
  int IKA, J, JC, JR;
  double ANORM, ULP, UNFL, WNORM;

  // Constants

  RESULT[1] = ZERO;
  RESULT[2] = ZERO;
  if (N <= 0) return;

  IKA = max(0, min(N - 1, KA));

  if (lsame(UPLO, 'U')) {
    LOWER = false;
    CUPLO = 'U';
  } else {
    LOWER = true;
    CUPLO = 'L';
  }

  UNFL = dlamch('Safe minimum');
  ULP = dlamch('Epsilon') * dlamch('Base');

  // Some Error Checks

  // Do Test 1

  // Norm of A:

  ANORM = max(zlanhb('1', CUPLO, N, IKA, A, LDA, RWORK), UNFL);

  // Compute error matrix:    Error = A - U S U**H

  // Copy A from SB to SP storage format.

  J = 0;
  for (JC = 1; JC <= N; JC++) {
    if (LOWER) {
      for (JR = 1; JR <= min(IKA + 1, N + 1 - JC); JR++) {
        J++;
        WORK[J] = A[JR][JC];
      }
      for (JR = IKA + 2; JR <= N + 1 - JC; JR++) {
        J++;
        WORK[J] = Complex.zero;
      }
    } else {
      for (JR = IKA + 2; JR <= JC; JR++) {
        J++;
        WORK[J] = Complex.zero;
      }
      for (JR = min(IKA, JC - 1); JR >= 0; JR--) {
        J++;
        WORK[J] = A[IKA + 1 - JR][JC];
      }
    }
  }

  for (J = 1; J <= N; J++) {
    zhpr(CUPLO, N, -D[J], U(1, J).asArray(), 1, WORK);
  }

  if (N > 1 && KS == 1) {
    for (J = 1; J <= N - 1; J++) {
      zhpr2(CUPLO, N, -E[J].toComplex(), U(1, J).asArray(), 1,
          U(1, J + 1).asArray(), 1, WORK);
    }
  }
  WNORM = zlanhp('1', CUPLO, N, WORK, RWORK);

  if (ANORM > WNORM) {
    RESULT[1] = (WNORM / ANORM) / (N * ULP);
  } else {
    if (ANORM < ONE) {
      RESULT[1] = (min(WNORM, N * ANORM) / ANORM) / (N * ULP);
    } else {
      RESULT[1] = min(WNORM / ANORM, N) / (N * ULP);
    }
  }

  // Do Test 2

  // Compute  U U**H - I

  zgemm('N', 'C', N, N, N, Complex.one, U, LDU, U, LDU, Complex.zero,
      WORK.asMatrix(), N);

  for (J = 1; J <= N; J++) {
    WORK[(N + 1) * (J - 1) + 1] -= Complex.one;
  }

  RESULT[2] = min(zlange('1', N, N, WORK.asMatrix(), N, RWORK), N) / (N * ULP);
}
